### PR TITLE
[feature-wip](array-type) add the array_join function

### LINKED ISSUE #10052: [Feature] Add lots of Array Functions
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

We have supported `array` data type recently, all pr(s) about `array` please refer to https://github.com/apache/incubator-doris/issues/7570 .

Now we need to add lots of array functions to make `array` more powerful.
If anybody is interesting to `array`, please let me know.

Array Functions DONE:
- size/cardinality
- array_position
- array_contains
- element_at

Array Functions DOING:
- array_sum
- array_avg
- array_min
- array_max
- array_product
- array_distinct
- array_remove
- arrays_overlap

Array Functions TODO:
- array_union
- array_intersect
- array_except
- reverse
- concat_ws
- array_join
- array_sort
- array_slice

### Use case

_No response_

### Related issues

https://github.com/apache/incubator-doris/issues/7570

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by carlvinhust2012**
# Proposed changes
Issue Number: close #10052 
1. this pr is used to add the array_join function.
2. the usage is as below:

```
mysql> set enable_array_type=true;
mysql> set enable_vectorized_engine=true;

mysql> select k1, k2, array_join(k1, '_', 'null') from array_test order by k1;
+------+-----------------------------+------------------------------------+
| k1   | k2                          | array_join(`k2`, '_', 'null')      |
+------+-----------------------------+------------------------------------+
|  1   | [1, 2, 3, 4, 5]             | 1_2_3_4_5                          |
|  2   | [6, 7, 8]                   | 6_7_8                              |
|  3   | []                          |                                    |
|  4   | NULL                        | NULL                               |
|  5   | [1, 2, 3, 4, 5, 4, 3, 2, 1] | 1_2_3_4_5_4_3_2_1                  |
|  6   | [1, 2, 3, NULL]             | 1_2_3_null                         |
|  7   | [4, 5, 6, NULL, NULL]       | 4_5_6_null_null                    |
+------+-----------------------------+------------------------------------+

mysql> select k1, k2, array_join(k2, '_', 'null') from array_test01 order by k1;
+------+-----------------------------------+------------------------------------+
| k1   | k2                                | array_join(`k2`, '_', 'null')      |
+------+-----------------------------------+------------------------------------+
|  1   | ['a', 'b', 'c', 'd']              | a_b_c_d                            |
|  2   | ['e', 'f', 'g', 'h']              | e_f_g_h                            |
|  3   | [NULL, 'a', NULL, 'b', NULL, 'c'] | null_a_null_b_null_c               |
|  4   | ['d', 'e', NULL, ' ']             | d_e_null_                          |
|  5   | [' ', NULL, 'f', 'g']             |  _null_f_g                         |
+------+-----------------------------------+------------------------------------+
```

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
4. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
5. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
6. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/array/function_array_join.h [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    static constexpr auto name = "array_join";
+    static FunctionPtr create() { return std::make_shared<FunctionArrayJoin>(); }
+    using NullMapType = PaddedPODArray<UInt8>;
+
+    /// Get function name.
+    String get_name() const override { return name; }
+
+    bool is_variadic() const override { return true; }
+
+    size_t get_number_of_arguments() const override { return 3; }
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_join.h:40]:
```suggestion
    size_t get_number_of_arguments() const override { return 0; }
```

File: be/src/vec/functions/array/function_array_join.h [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+        const ColumnType* src_data_concrete = reinterpret_cast<const ColumnType*>(&src_column);
+        if (!src_data_concrete) {
+            return false;
+        }
+
+        ColumnArray::Offset prev_src_offset = 0;
+        for (auto curr_src_offset : src_offsets) {
+            std::string result_str;
+            std::vector<std::string> str_vec;
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_join.h:146]: append result directly, no need to use str_vec

**Comment by HappenLee** [comment]
the function is a agg function. we should use group_concat to do the work. like `sum` `avg` `max`

**Comment by carlvinhust2012** [comment]
> the function is a agg function. we should use group_concat to do the work. like `sum` `avg` `max`
We consider the compatibility with the hdfs ecosystem and follow the spark syntax.
https://spark.apache.org/docs/latest/sql-ref-functions-builtin.html

File: be/src/vec/functions/array/function_array_join.h (hunk 2 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_join.h]: done

File: be/src/vec/functions/array/function_array_join.h (hunk 1 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_join.h]: done

**Review by xy720 — APPROVED** [review]
LGTM

**Comment by cambyzju** [comment]
LGTM

**Pull request merged**